### PR TITLE
Allow bandwise slicing with non-default step.  Fixes #308

### LINKED ISSUE #308: step value for bandwise slice is silently ignored
It looks like the syntactic sugar of using `__getitem__`  with a slice to get at bands on  `pyvips.Image` objects is silently ignoring the step part of the slice.

For example:
```python
import pyvips

xy = pyvips.Image.xyz(3, 4)
xyxyxy = xy.bandjoin(xy).bandjoin(xy)

print(f'OK:  {xy[0].max() = }')
print(f'OK:  {xy[1].max() = }')
print(f'OK:  {xyxyxy.bands = }')
print(f'OK:  {xyxyxy[:3].bands = }')
print(f'BUG: {xyxyxy[::2].bands = }')  # step silently ignored
# reversing x,y into row,col via idiomatic [::-1] reversing slice
print(f'BUG: {(xy[::-1] == xy).min() = }')  # These match because step=-1 is silently ignored

# OK:  xy[0].max() = 2.0
# OK:  xy[1].max() = 3.0
# OK:  xyxyxy.bands = 6
# OK:  xyxyxy[:3].bands = 3
# BUG: xyxyxy[::2].bands = 6
# BUG: (xy[::-1] == xy).min() = 255.0

```

Thanks in advance!

### PR DESCRIPTION
**Pull request by erdmann**
Hi @jcupitt,

Here is a small 3-line pull request to fix #308.

--Rob

### CODE REVIEW

**Comment by erdmann** [comment]
By reading your code I just learned that it was possible to do `im.bandjoin(bands)`, thereby obviating the need for my `functools.reduce(lambda a, b: a.bandjoin(b), bands)` approach.   Good to know!  

On the other hand, the whole section on slicing could be replaced by just these two lines:

```python
bands = [ self.extract_band(i) for i in range(self.bands)[arg] ]
return functools.reduce(lambda u, v: u.bandjoin(v), bands)
```

Are there performance implications for taking contiguous bands, splitting them out like this, and subsequently bandjoining them together in the case in which the step is 1?


Note that I also added a few extra lines to the documentation about these fancy slices.

**Comment by jcupitt** [comment]
Here's the start of some tests for this as a git diff:

```
diff --git a/tests/test_constants.py b/tests/test_constants.py
index fc1d5a7..92727fa 100644
--- a/tests/test_constants.py
+++ b/tests/test_constants.py
@@ -15,12 +15,10 @@ class TestConstants:
         assert im.height == 1
 
     def test_array_const_args(self):
-        r = pyvips.Image.black(16, 16)
-        r = r.draw_rect(255, 10, 12, 1, 1)
-        g = pyvips.Image.black(16, 16)
-        g = g.draw_rect(255, 10, 11, 1, 1)
-        b = pyvips.Image.black(16, 16)
-        b = b.draw_rect(255, 10, 10, 1, 1)
+        black = pyvips.Image.black(16, 16)
+        r = black.draw_rect(255, 10, 12, 1, 1)
+        g = black.draw_rect(255, 10, 11, 1, 1)
+        b = black.draw_rect(255, 10, 10, 1, 1)
         im = r.bandjoin([g, b])
 
         assert im.width == 16
diff --git a/tests/test_image.py b/tests/test_image.py
index 675ec9d..5605774 100644
--- a/tests/test_image.py
+++ b/tests/test_image.py
@@ -16,3 +16,36 @@ class TestImage:
         many_page = image.pagejoin([image, image])
         image_list = many_page.pagesplit()
         assert len(image_list) == 3
+
+    def test_bandjoin(self):
+        black = pyvips.Image.black(16, 16)
+        a = black.draw_rect(1, 0, 0, 1, 1)
+        b = black.draw_rect(2, 0, 0, 1, 1)
+        c = black.draw_rect(3, 0, 0, 1, 1)
+        im = a.bandjoin([b, c, a, b, c, a, b, c])
+
+        assert im.width == 16
+        assert im.height == 16
+        assert im.bands == 9
+
+    def test_bandslice(self):
+        black = pyvips.Image.black(16, 16)
+        a = black.draw_rect(1, 0, 0, 1, 1)
+        b = black.draw_rect(2, 0, 0, 1, 1)
+        c = black.draw_rect(3, 0, 0, 1, 1)
+        im = a.bandjoin([b, c, a, b, c, a, b, c])
+
+        x = im[1]
+
+        assert x.bands == 1
+        assert x(0, 0) == [2]
+
+        x = im[1:4]
+
+        assert x.bands == 3
+        assert x(0, 0) == [2, 3, 1]
+
+        x = im[4:1]
+
+        assert x.bands == 3
+        assert x(0, 0) == [1, 2, 3]
```

File: pyvips/vimage.py [hunk 1]
```diff
@@ -998,10 +998,17 @@ def __getitem__(self, arg):
...
+            every_other = multiband[::2]
+            other_every_other = multiband[1::2]
 
         """
 
         if isinstance(arg, slice):
+            if arg.step is not None: # fancy slicing, though this would work for any slice
+                bands = [ self.extract_band(i) for i in range(self.bands)[arg] ]
+                return bands[0].bandjoin(bands[1:])
+
```
Comment (jcupitt) [inline pyvips/vimage.py:1011]: I try to avoid early returns if possible, except for error checks.

Comment (erdmann) [reply pyvips/vimage.py:1018]: Latest implementation addresses this.

**Comment by erdmann** [comment]
Ok, I wrote what I think is a simple, efficient, and robust `__getitem__`, and also added a slew of tests for all the edge cases I could think of (thanks for the testing starter!).

In all cases, I'm following the principle of least astonishment by adhering exactly to the semantics of indexing/slicing integer ranges in Python.   In other words, the indices of the band or bands you get correspond exactly to indexing/slicing `range(self.bands)`.

I ran `pytest tests` in the pyvips root and all the new tests pass.

I know my past contributions have just been raising issues; thanks for the opportunity to make this wee code contribution to such a great library.

File: pyvips/vimage.py [hunk 2]
```diff
@@ -998,33 +998,35 @@ def __getitem__(self, arg):
...
+            reversed  = multiband[::-1]
+            every_other = multiband[::2]
+            other_every_other = multiband[1::2]
+
+        In all cases, the semantics of slicing exactly match those of slicing 
+        `range(self.bands)`.
 
         """
 
+        band_range = range(self.bands)[arg] # raises IndexError for bad integer indices
```
Comment (erdmann) [inline pyvips/vimage.py:1010]: slicing a range object with an in-range integer yields an integer, while slicing it with a slice keeps it as a range object, so we let this do all the work of keeping track of start, stop, step, len, while also using the fact that bad integer slices will already raise an out-of-range `IndexError`.

**Comment by jcupitt** [comment]
That looks great! Thank you for doing this work. As you say, it's nice and simple now.

Could you add a line to the CHANGELOG as well? Plus a credit to yourself, of course.

File: pyvips/vimage.py (hunk 2 above)
Comment (jcupitt) [reply pyvips/vimage.py:1010]: I didn't know about this feature. Neat!

**Comment by erdmann** [comment]
Ok, I added an entry to the CHANGELOG.

**Comment by jcupitt** [comment]
Nice, let's merge. Thanks again!

**Comment by erdmann** [comment]
Thanks!  I noticed that you moved the CHANGELOG entry out from under the master branch heading and into the "## Version 2.1.17 (15 Dec 2021)" section.   Was this intentional?

**Comment by jcupitt** [comment]
Yes, 2,1,16 is the currently released version, 2.1.17 is the one in development right now.

**Comment by erdmann** [comment]
Ah, ok.  I read it as saying that 2.1.17 was released 15 Dec 2021, especially since there was also a "master" heading at the top where I inferred one should put not-yet-released features.
